### DISCUSSION TITLE
Can I use async and sync lock on same AsyncReaderWriterLock ?

### DISCUSSION THREAD

**Discussion by oruchreis** [timcassell/ProtoPromise#441] [Q&A] [answered]
Hi,
I saw both sync and async methods on `AsyncReaderWriterLock`. Can I use same lock object in both sync and async methods? I was using DotNext's `AsyncReaderWriterLock`, and I've figured out it doesn't wait if you call WriteLockAsync in a thread and then call ReadLock in another thread, also [that question](https://github.com/dotnet/dotNext/discussions/237) explains much more.
I've tested in a sample project and it works, but I want to ask and be sure if it works.
```csharp
private static readonly AsyncReaderWriterLock _rwlock = new();
//Thread:1 with async using
using (await _rwlock.WriterLockAsync(token))
{}
//Thread:2 with sync using
using (_rwlock.UpgradeableReaderLock())
{
    using(_rwlock.WriterLock())
    {}
}
```

**Comment by timcassell** [comment, marked as answer]
Hi @oruchreis, sorry I didn't see this until now (I usually get notifications on this repo, not sure why I didn't for this discussion).

Yes, `AsyncReaderWriterLock` is designed to work with both async and sync on the same instance. It's not really recommended due to deadlock issues, but if you know for a fact that the sync calls will only happen on their own thread, that's exactly the situation it was designed to support.

**Comment by timcassell** [comment]
Btw, your upgradeable lock example is used incorrectly. It should look like this:

```cs
//Thread:2 with sync using
using (var upgradeableReaderKey = _rwlock.UpgradeableReaderLock())
{
    using(_rwlock.UpgradeToWriterLock(upgradeableReaderKey))
    {}
}
```

Otherwise, your normal `_rwlock.WriterLock()` causes a deadlock with the `_rwlock.UpgradeableReaderLock()`.
